### PR TITLE
(CDAP-7166) Added an admin manual for secure storage and an integrati…

### PR DESCRIPTION
**Pull request by bdmogal**
…ons manual for Apache Hadoop KMS.

Build: http://builds.cask.co/browse/CDAP-DQB96-2
Relevant pages: 
[Admin Manual](http://builds.cask.co/artifact/CDAP-DQB96/shared/build-2/Docs-HTML/3.5.0/en/admin-manual/security/secure-storage.html)
[Integrations - Hadoop KMS](http://builds.cask.co/artifact/CDAP-DQB96/shared/build-2/Docs-HTML/3.5.0/en/integrations/hadoop-kms.html)

### CODE REVIEW

File: cdap-docs/admin-manual/source/security/secure-storage.rst [hunk 1]
```diff
@@ -4,12 +4,48 @@
...
 
-==========================
-Configuring Secure Storage
-==========================
+==============
+Secure Storage
+==============
 
-**Secure Storage** allows users to store sensitive information such as passwords in an encrypted manner.
+Applications may need access to sensitive data such as *passphrases*, *cryptographic keys*, *access tokens* or
```
Comment (Nishith) [inline cdap-docs/admin-manual/source/security/secure-storage.rst]: comma before "or"?

File: cdap-docs/admin-manual/source/security/secure-storage.rst [hunk 2]
```diff
@@ -4,12 +4,48 @@
...
 
 **Note:** In CDAP 3.5.0, encryption and decryption of the contents only happens at the
 secure store, not while the data is transitting to the secure store. In a later version of
 CDAP, all transport involving secure keys will be secured using SSL.
+
+Secure Storage Format
+---------------------
+An entry in secure storage consists of:
+
+- **Key**: A alias for the entry, also referred to in this document and elsewhere as :term:`Secure Key <secure key>`.
```
Comment (Nishith) [inline cdap-docs/admin-manual/source/security/secure-storage.rst]: "An alias"?

File: cdap-docs/admin-manual/source/security/secure-storage.rst [hunk 3]
```diff
@@ -4,12 +4,48 @@
...
+CDAP provides two different implementations of secure storage, depending on the runtime:
+
+File-backed Secure Storage
+--------------------------
+File-based secure storage is available for use in in-memory (unit-test) and standalone (SDK). It uses a
+`Java JCEKS <http://docs.oracle.com/javase/7/docs/technotes/guides/security/crypto/CryptoSpec.html#KeyManagement>`__
+based implementation for storing secure keys. This implementation is not available in
+:term:`Distributed CDAP <distributed cdap>`, since it stores secure data on the local file system, and will therefore
+not be available on all nodes of a distributed cluster. To use this mode, please set ``security.store.provider`` to
+``file`` in ``cdap-site.xml``.
```
Comment (Nishith) [inline cdap-docs/admin-manual/source/security/secure-storage.rst]: Should also mention "security.store.file.password"?

File: cdap-docs/admin-manual/source/security/secure-storage.rst [hunk 4]
```diff
@@ -4,12 +4,49 @@
...
+Secure Storage Format
+---------------------
+An entry in secure storage consists of:
+
+- **Key**: An alias for the entry, also referred to as a :term:`Secure Key <secure key>`.
+  Data is stored against the provided key and can be retrieved using the same key. 
+  Key must be of the :ref:`Alphanumeric Character Set <supported-characters>`, contain *only
+  lowercase* characters, and should start with a letter.
+- **Data**: The data which is to be stored in a secure an encrypted manner. This could be a passphrase, 
+  cryptographic key, access token, or any other data that needs to be stored securely.
```
Comment (awholegunch) [inline cdap-docs/admin-manual/source/security/secure-storage.rst:31]: Is there a size limit on this?

File: cdap-docs/admin-manual/source/security/secure-storage.rst [hunk 5]
```diff
@@ -4,12 +4,49 @@
...
 
 **Note:** In CDAP 3.5.0, encryption and decryption of the contents only happens at the
 secure store, not while the data is transitting to the secure store. In a later version of
 CDAP, all transport involving secure keys will be secured using SSL.
+
+Secure Storage Format
+---------------------
+An entry in secure storage consists of:
+
+- **Key**: An alias for the entry, also referred to as a :term:`Secure Key <secure key>`.
```
Comment (awholegunch) [inline cdap-docs/admin-manual/source/security/secure-storage.rst:26]: Is there a key length limit?

File: cdap-docs/hydrator-manual/source/glossary.rst [hunk 6]
```diff
@@ -95,6 +95,10 @@ Cask Hydrator Glossary
       A type of :term:`Hydrator Pipeline` that runs continuously, performing actions on
       a distinct set of data.
 
+   Secure Key
+      A key for an entry in :ref:`Secure Storage <admin-secure-storage>`. An entry in secure
+      storage can be referenced and retrieved using a Secure Key.
+
```
Comment (awholegunch) [inline cdap-docs/hydrator-manual/source/glossary.rst]: This is not the correct file for this entry. It should instead be added to the main glossary, located in `reference-manual/source/glossary.rst`

File: cdap-docs/integrations/source/hadoop-kms.rst [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    :author: Cask Data, Inc.
+    :copyright: Copyright © 2016 Cask Data, Inc.
+
+.. _apache-hadoop-kms:
+
+=========================================
+Apache Hadoop Key Management Server (KMS)
+=========================================
+CDAP integrates with Apache Hadoop KMS as the backend for :ref:`Secure Storage <admin-secure-storage>`. To use this
+secure storage implementation, please set ``security.store.provider`` to ``kms`` in ``cdap-site.xml``.
```
Comment (awholegunch) [inline cdap-docs/integrations/source/hadoop-kms.rst]:
```
secure storage implementation, set ``security.store.provider`` to ``kms`` in ``cdap-site.xml``.

File: cdap-docs/admin-manual/source/security/secure-storage.rst (hunk 4 above)
Comment (Nishith) [reply cdap-docs/admin-manual/source/security/secure-storage.rst:31]: I could not find a size limitation mentioned any where for either KMS or the java keystore.

File: cdap-docs/admin-manual/source/security/secure-storage.rst (hunk 5 above)
Comment (Nishith) [reply cdap-docs/admin-manual/source/security/secure-storage.rst:26]: I could not find a size limitation mentioned any where for either KMS or the java keystore.

**Comment by awholegunch** [comment]
Though the glossary needs adjusting, I have kicked off another build: http://builds.cask.co/browse/CDAP-DQB96-5

File: cdap-docs/hydrator-manual/source/glossary.rst (hunk 6 above)
Comment (bdmogal) [reply cdap-docs/hydrator-manual/source/glossary.rst]: oops! yes, fixed.

**Comment by bdmogal** [comment]
@awholegunch @Nishith addressed comments. Please review again.

**Comment by Nishith** [comment]
👍 LGTM.

File: cdap-docs/reference-manual/source/glossary.rst [hunk 8]
```diff
@@ -49,13 +49,13 @@ Glossary
...
-      storage can be referenced and retrieved using a Secure Key via :ref:`Programmatic <secure-keys-programmatic>`
-      or :ref:`RESTful <http-restful-api-secure-storage>` API.
+      An identifier or an alias for an entry in :term:`Secure Storage`. An entry in secure
+      storage can be referenced and retrieved using a Secure Key using :ref:`programmatic <secure-keys-programmatic>`
+      or :ref:`RESTful <http-restful-api-secure-storage>` APIs.
 
    Secure Storage
-      Storage for :term:`Secure Key`. CDAP supports :ref:`File Backed <admin-secure-storage-file>`
-      (for :term:`Standalone CDAP`) as well as :ref:`Apache Hadoop KMS Backed <admin-secure-storage-kms>`
+      Storage using a :term:`Secure Key`. CDAP supports :ref:`File-backed <admin-secure-storage-file>`
```
Comment (bdmogal) [inline cdap-docs/reference-manual/source/glossary.rst]: this is storage for secure keys, not using secure keys, no?

Comment (Nishith) [reply cdap-docs/reference-manual/source/glossary.rst]: Storage for sensitive data using a key?

**Comment by awholegunch** [comment]
LGTM

**Comment by bdmogal** [comment]
Thanks @awholegunch @Nishith! Squashed commits. Running a new build - http://builds.cask.co/browse/CDAP-DQB96-8. Will merge after the build succeeds.

**Comment by bdmogal** [comment]
Build succeeded - http://builds.cask.co/browse/CDAP-DQB96-8. Merging.

**Pull request merged**